sno/working_copy.py

```diff
@@ -1024,78 +1020,98 @@ def _apply_meta_schema(self, src_ds, dest_ds, src_value, dest_value, dbcur):
-                    # GIT_DELTA_UNTRACKED
-                    raise NotImplementedError(f"Delta status: {d.status_char()}")
-
-                src_value = src_ds.get_meta_item(path)
-                dest_value = dest_ds.get_meta_item(path)
-                func(src_ds, dest_ds, src_value, dest_value, dbcur)
-
-    def _apply_feature_diff(self, src_ds, dest_ds, dbcur, feature_diff_index):
+
+    _INSERT_UPDATE_DELETE = (
+        pygit2.GIT_DELTA_ADDED,
+        pygit2.GIT_DELTA_MODIFIED,
+        pygit2.GIT_DELTA_DELETED,
+    )
+
+    def _apply_feature_diff(
+        self, base_ds, target_ds, dbcur, track_changes_as_dirty=False
+    ):
+        """
+        Change the features of this working copy from their current state, base_ds - to the desired state, target_ds.
+        base_ds - dataset containing the features that match the WC table currently.
+        target_ds - dataset containing the desired features of the WC table.
+        dbcur - DB cursor.
+        track_changes_as_dirty - whether to track these changes as working-copy edits in the tracking table.
+        """
+        feature_diff_index = base_ds.feature_tree.diff_to_tree(target_ds.feature_tree)
+        if not feature_diff_index:
+            return
+
         L.debug("Feature diff: %s changes", len(feature_diff_index))
-        for d in feature_diff_index.deltas:
-            # TODO: improve this by grouping by status then calling
-            # write_features/delete_features passing multiple PKs?
-            if d.status == pygit2.GIT_DELTA_DELETED:
-                old_pk = src_ds.decode_path_to_1pk(os.path.basename(d.old_file.path))
-                L.debug("reset(): D %s (%s)", d.old_file.path, old_pk)
-                self.delete_features(dbcur, src_ds, [old_pk])
-            elif d.status == pygit2.GIT_DELTA_MODIFIED:
-                old_pk = src_ds.decode_path_to_1pk(os.path.basename(d.old_file.path))
-                new_pk = dest_ds.decode_path_to_1pk(os.path.basename(d.new_file.path))
-                L.debug(
-                    "reset(): M %s (%s) -> %s (%s)",
-                    d.old_file.path,
-                    old_pk,
-                    d.new_file.path,
-                    new_pk,
-                )
-                self.write_features(dbcur, dest_ds, [new_pk])
-            elif d.status == pygit2.GIT_DELTA_ADDED:
-                new_pk = dest_ds.decode_path_to_1pk(os.path.basename(d.new_file.path))
-                L.debug("reset(): A %s (%s)", d.new_file.path, new_pk)
-                self.write_features(dbcur, dest_ds, [new_pk])
-            else:
-                # GIT_DELTA_RENAMED
-                # GIT_DELTA_COPIED
-                # GIT_DELTA_IGNORED
-                # GIT_DELTA_TYPECHANGE
-                # GIT_DELTA_UNMODIFIED
-                # GIT_DELTA_UNREADABLE
-                # GIT_DELTA_UNTRACKED
-                raise NotImplementedError(f"Delta status: {d.status_char()}")
+
+        first_unsupported_delta = next(
```

Comment:
This change means we're iterating the entire diff 3 times.

I'd prefer to not do that - it could be millions of features...

what about using a single iteration? Maybe use a chunk iterator to batch the deletes/inserts to sqlite every 10000 deltas or something.

And if we encounter an unsupported delta we can rollback the sqlite transaction before raising

Reply:
Done - appended all the PKS to lists before resetting them. There doesn't seem to be any real downside to doing it that way - allocating a million element list itself is very cheap compared to generating a million-element diff, which is what we did at the top of this method.